torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -886,68 +929,122 @@ void ProcessGroupNCCL::logWorkEnd(WorkNCCL& work) {
+
+        // If work hits an exception (either an error or timeout)
+        if (work.exception()) {
+          if (SHOULD_CLEAN_UP(asyncErrorHandling_)) {
+            // Abort work and corresponding communicators
+            work.abort();
+            // PG level abort, which would abort all other communicators on this
+            // rank
+            abort();
+          }
+          // Report desync state in case of timeout
+          if (desyncDebug_ && timedOut) {
+            try {
+              auto desyncMsg =
+                  retrieveDesyncReport(store_, "NCCL", rank_, size_);
+              LOG(ERROR) << desyncMsg;
+            } catch (const std::exception& e) {
+              LOG(ERROR) << "Failed to retrieve NCCL_DESYNC_DEBUG report. "
+                         << " Please file an issue. Error: " << e.what();
+            } catch (...) {
+              LOG(ERROR)
+                  << "Failed to rerieve NCCL_DESYNC_DEBUG report with unknown error."
+                  << " Please file an issue.";
+            }
           }
+          // Throw exception
+          work.handleException(asyncErrorHandling_);
         }
-        // Throw exception
-        work.handleException(asyncErrorHandling_);
-      }
 
-      // Work status logging for desync debug
-      if (desyncDebug_) {
-        if (work.isStarted()) {
-          logWorkStart(work);
+        // Work status logging for desync debug
+        if (desyncDebug_) {
+          if (work.isStarted()) {
+            logWorkStart(work);
+          }
+          if (work.isCompleted()) {
+            logWorkEnd(work);
+          }
         }
+
+        // Clean up completed work
         if (work.isCompleted()) {
-          logWorkEnd(work);
+          if (onCompletionHook_) {
+            // onCompletion hook will consume the work, so keep them alive in a
+            // separate list for now
+            completedWorkList.splice(
+                completedWorkList.begin(), workMetaList_, it++);
+          } else {
+            it = workMetaList_.erase(it);
+          }
+        } else {
+          // Increment the iterator if the current WorkNCCL object is not
+          // completed.
+          ++it;
         }
       }
+      done = workMetaList_.empty();
+    }
+    // Python hook will grab GIL. Release workMetaListMutex_ before acquiring
+    // GIL to avoid potential deadlock.
 
-      // Clean up completed work
-      if (work.isCompleted()) {
-        it = workMetaList_.erase(it);
-      } else {
-        // Increment the iterator if the current WorkNCCL object is not
-        // completed.
-        ++it;
+    try {
+      while (!completedWorkList.empty()) {
+        // The work object will be passed to Python side. Wrap it
+        // with an intrusive_ptr so that pybind can safely pass by copy
+        // and own the ptr lifetime.
+        onCompletionHook_(
+            c10::make_intrusive<WorkNCCL>(completedWorkList.back()));
```

Comment:
Why are we copying the Work object?

Reply:
`Work` does not have move ctor. Let me know if you want me to add that.